### DISCUSSION TITLE
app.read_property getting -no object class- but running the module doesn't have the problem

### DISCUSSION THREAD

**Discussion by jerrys987** [JoelBender/bacpypes#539] [General]
I am just starting out with bacpypes3, so sorry if this is well understood.
I started with discover-devices.py and modified it to do some different things to try to get a handle on how to work with the module.  I looked at the module logic that is returning the message and am not sure why it isn't working. When I use the module shell, it works fine.
python3 -m bacpypes3 --address 172.18.1.29/24
> read 172.18.1.210 analog-input,0 present-value
72.80000305175781

so there is clearly something I am missing.

code block:
        for i_am in i_ams:
            if _debug:
                _log.debug("    - i_am: %r", i_am)

            device_address: Address = i_am.pduSource
            device_identifier: ObjectIdentifier = i_am.iAmDeviceIdentifier
            print(f"{device_identifier} @ {device_address}")

            try:
                device_name: str = await app.read_property(
                    device_address, device_identifier, "object-name"
                )
                print(f"    name: {device_name}")
                model_name: str = await app.read_property(
                    device_address, device_identifier, "model-name"
                )
                print(f"    model: {model_name}")
                temp = await app.read_property(device_address, 'analog-input,0', 'present-value')
                print(str(temp))

output:
device,2002 @ 172.18.1.210
    name: thermostat 2
    model: 122C
-no object class-
device,2003 @ 172.18.1.211
    name: thermostat 1
    model: 122C
-no object class-

hope someone can point me in the right direction.

**Comment by jerrys987** [comment]
If I follow the exact pattern that the command shell uses 
get_device_info, get_vendor_info, vendor_info.object_identifier('analog_input,0'), PropertyReference, app.read_property()
the value comes through correctly.
I am running python3.12.3 on ubuntu24.04

this seems really messy

**Comment by JoelBender** [comment]
Most of the time you will be reading and writing standard objects (like `analog-input`) and standard properties (like `present-value`).  However, BACnet supports vendor proprietary properties to standard objects and proprietary objects.  Translating the string `snork` into the correct enumeration value requires knowing the vendor of the device (by using the vendor identifier value from the device object), some "schema" for the name to enumeration mapping for the request (a subclass of `PropertyIdentifier`), a similar mapping from the `my-special-object` object type name to its "schema" for the object (a subclass of `Object`) which _then_ provides decoding logic.

Run your application again with a `--debug bacpypes3.service.object.ReadWritePropertyServices` and look at the contents of the `propertyValue` tag list.  There will be an opening tag, some content that could be application or context encoded, and a closing tag.  You will also see the `device_info` returned from the cache (which may be `None` which implies there is no custom schema for that vendor) and the `vendor_info` object used for decoding (which falls back to ASHRAE).

**Reply by jerrys987** [reply]
Thanks for the explanation. What was/is not obvious to me is why I can do a read_property on device (app.read_property(dev_address, dev_id, 'model-name')) without having to go through the process of getting the the device_info and vendor_info but have to do that for 'analog-input, 0','present-value' ?

**Comment by JoelBender** [comment]
Yea, I get that, and I don't know.  Little comfort that this is the only bug like this reported and that I can't reproduce it.  So go to the mattresses; turn on that debugging option and let me know what it says, and if that doesn't yield any clues, run Wireshark and post a hex dump (which reminds me it would be nice to have some kind of "drop box" for people to send me pcap files privately or obfuscate network addresses and such).

**Comment by jerrys987** [comment]
I have run it with debug and have a pcap as well. All are attached.
[bpdebug.zip](https://github.com/user-attachments/files/17787982/bpdebug.zip)

**Comment by JoelBender** [comment]
The successful times you are calling it with an `ObjectIdentifier`:
```
DEBUG:bacpypes3.service.object.ReadWritePropertyServices:read_property <IPv4Address 172.18.1.210> (<ObjectType: device>, 2002) 'object-name' None
```
The unsuccessful ones you are calling it with a string:
```
DEBUG:bacpypes3.service.object.ReadWritePropertyServices:read_property <IPv4Address 172.18.1.210> 'analog-input,0' 'present-value' None
```
The code doesn't assert that the parameter is actually an instance of an `ObjectIdentifier` so the call `vendor_info.get_object_class(objid[0])` is looking for an `'a'` which it is not finding.  For the short term please pass `ObjectIdentifer('analog-input,0')` as the parameter.

It would be better if it worked as you are using it (the principle of [least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)) and that goes for the `prop` as well.  I'm thing about changing the function interface to this:
```
    async def read_property(
        self,
        address: Address,
        objid: Union[ObjectIdentifier, str],
        prop: Union[PropertyIdentifier, str],
        array_index: Optional[int] = None,
    ) -> _Any:
        ...
```
and adding some kind of `def resolve_zzz(self, ...)` to the application that can be used with other methods like `write_property()`, etc.

**Comment by jerrys987** [comment]
Thanks you for such prompt work on this problem.
I will make the change and test again.
jerry

**Comment by jerrys987** [comment]
works fine now. Should I also change the present-value to PropertyIdentifier('present-value')?

**Comment by JoelBender** [comment]
There is an issue [here](https://github.com/JoelBender/BACpypes3/issues/62) with a [branch](https://github.com/JoelBender/BACpypes3/tree/62-string-parameters-to-read_property-and-write_property) for this feature.  This is a pretty significant refactoring, please test it when you get a chance.  I'll make sure it doesn't break BAC0 before merging it.

**Comment by jerrys987** [comment]
Sorry, it took a few days to get back to this. I checked the branch and it works fine without the typecast as well as with it now. You like nice short branch names (my poor fish command string blew out.) :)
